### PR TITLE
SHT45 interface

### PR DESCRIPTION
**Pull request by matias-cphatomics**
Implementation of SHT45 humidity sensor interface. 

**Note:** The SPI communication is setup in a way where a command is first sent to make the sensor perform a corresponding action e.g. there are three types of measurement accuracies which takes has three different acquisition times. As of now these wait times are hard-coded as HAL_Delays. It would also be possible to store the time of sending the command and then let the user make the SPI getting call whenever a sufficient time has elapsed.

The upside of using HAL_Delays is that the user does not need to communicate with the sensor multiple times to get a measurement. The downside is of course that it stalls the program. 

@LukeWalker-CA let me know if you find the use of HAL_Delays bad in which case I will convert it to a state-machine-ish implementation.

### CODE REVIEW

File: STM32/I2C/Src/sht45.c [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+{
+    dev->data.absolute_humidity = 6.112 * exp((17.67*dev->data.temperature)/(dev->data.temperature+243.5))*dev->data.relative_humidity*2.1674/(273.15+dev->data.temperature);
+}
+
+/*!
+** @brief Function to send any valid command to SHT4x chip (except for abort call)
+*/
+static HAL_StatusTypeDef sht4x_set_mode(sht4x_handle_t *dev, uint8_t command)
+{
+    if (HAL_I2C_Master_Transmit(dev->hi2c, dev->device_address << 1u, &command, 1, 10) != HAL_OK) {
```
Comment (LukeWalker-CA) [inline STM32/I2C/Src/sht45.c:63]: 10 ms to send 8-bits seems a little high

File: STM32/I2C/Src/sht45.c [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+}
+
+/*!
+** @brief Get temperature and RH% measurement
+*/
+HAL_StatusTypeDef sht4x_get_measurement(sht4x_handle_t *dev, uint8_t command)
+{
+    HAL_StatusTypeDef ret = HAL_ERROR;
+    // Send measurement command to output temp and humidity from SHT45 at next read 
+    ret = sht4x_set_measurement_mode(dev, command);
```
Comment (LukeWalker-CA) [inline STM32/I2C/Src/sht45.c:162]: The above two lines can be combined, since the return value is always set

**Review by LukeWalker-CA — APPROVED** [review]
Not very controversial. 

There _are_ a lot of HAL_Delay()s, its not completely magic. The humidity board doesn't have strict timing right though, so it shouldn't matter too much

File: STM32/I2C/Inc/sht45.h [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+/***************************************************************************************************
+** PUBLIC FUNCTION DECLARATIONS
+***************************************************************************************************/
+
+HAL_StatusTypeDef sht4x_soft_reset(sht4x_handle_t *dev);
+HAL_StatusTypeDef sht4x_abort_command(sht4x_handle_t *dev);
+HAL_StatusTypeDef sht4x_get_serial(sht4x_handle_t *dev);
+HAL_StatusTypeDef sht4x_get_measurement(sht4x_handle_t *dev, uint8_t command);
+HAL_StatusTypeDef sht4x_turn_on_heater(sht4x_handle_t *dev, uint8_t heating_program);
```
[the remark this answers is not in the document]
Comment (matias-cphatomics) [reply STM32/I2C/Inc/sht45.h:66]: We can do it, but as opposed to the SHT35 there is no logical call to make as a part of the init. I ended up just calling the get_serial in the humidity PR. 

I'm happy to do some assertions or the like and perhaps calling the get_serial from within a init function if you prefer.

**Review by LukeWalker-CA — APPROVED** [review]

File: STM32/I2C/Src/sht45.c [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+ * 
+ *  Datasheet: https://sensirion.com/media/documents/33FD6951/662A593A/HT_DS_Datasheet_SHT4x.pdf
+ */
+
+#include "sht45.h"
+#include <math.h>
+
+/***************************************************************************************************
+** PRIVATE FUNCTIONS
+***************************************************************************************************/
```
[the remark this answers is not in the document]
Comment (matias-cphatomics) [reply STM32/I2C/Src/sht45.c:15]: I'm happy to put them there, but why do you think it would be good? :=)

**Pull request merged**